revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,180 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a certificate chain
+// as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ocsp"
+)
+
+type Options struct {
+	CertChain       []*x509.Certificate
+	SigningTime     time.Time
+	HTTPClient      *http.Client
+	MergeErrorsFunc func(errors []error) error
+}
+
+const pkixNoCheckOID string = "1.3.6.1.5.5.7.48.1.5"
+const ocspMaxResponseSize int64 = 20480 //bytes
+
+func OCSPStatus(opts Options) error {
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")}
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")}
+			}
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	certResults := make([]error, len(opts.CertChain))
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			wg.Add(1)
+			// Assume cert chain is accurate and next cert in chain is the issuer
+			go certOCSPStatus(cert, opts.CertChain[i+1], &certResults[i], &wg, opts)
+		} else {
+			// Last is root cert, which will never be revoked by OCSP
+			certResults[len(opts.CertChain)-1] = nil
+		}
+	}
+
+	wg.Wait()
+	return opts.MergeErrorsFunc(certResults)
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert *x509.Certificate, issuer *x509.Certificate, result *error, wg *sync.WaitGroup, opts Options) {
```

Comment:
1. Why are we taking error as input and not returning it ?
2. @shizhMSFT @patrickzheng200 can you please review code for parallelizing the ocsp calls? Passing wg just to call `wg.Done` looks odd but I am not  much familiar with concurrency in golang.

Reply:
1. We don't return it because this function is called in a goroutine. Rather than return the error, we update the reference to it so that it can be done asynchronously. This is essentially the same as returning, as we'll have access to all the errors once the goroutines are done.
2. Passing wg was a suggestion by @shizhMSFT in order to prevent race conditions. We generate a new wg for each unique call to OCSPStatus, which is passed to certOCSPStatus to decrement the correct counter once it is done.